### DISCUSSION TITLE
Zenith angle definition and post-processing of BRF

### DISCUSSION THREAD

**Discussion by Coder-GAN** [eradiate/eradiate#424] [Q&A] [answered]
Hello there, thanks so much for developing such a wonderful model! I have some small questions when I use the eradiate.

1. what`s the specific definition of zenith angle in eradicate? Is the zenith angle range from 0° to 90° or 0° to 180°?

2. I am using Eradiate to simulate the Top of Canopy BRF, the specific step was using 'MultiDistantMeasure.from_angles' coupled with 'experiments.CanopyExperiment'. Since I want to compare the BRF with Sentinel-2 L2A products (B2-B8A), I set the [srf = 'sentinel_2a-msi-2-raw'], the simulated results in each band to provide the multiple  BRF (corresponding to the wavelength range of S2 bands), but the Sentinel-2 L2A products only give a single BRF in each band. thus the question is how to process the simulated results into a single BRF in each band to be compared with the S2 L2A product.

Thanks very much for your attention!

**Comment by Coder-GAN** [comment]
The below contents were the updated answer from Vincent:

1.	The zenith angle is the angle between the local normal to the surface at a reference point, and a direction vector oriented outwards the reference point. 0° means zenith, 90° means horizon. An value higher than 90° usually makes little sense in the context of a planar reference geometry (e.g. under a plane-parallel atmosphere).
A notable special case is when we represent values in the principal plane: our output has a zenith coordinate ranging from -90° to +90°. This is a convention, and negative values correspond to an azimuth offset of +180°.

2.	I am not sure if I understand your question. If you set the SRF to that of Sentinel-2 band 2, then you should get a value for band 2 only. If you want to simulate bands 2 to 8A, you should define multiple measures, each with a different SRF. Does this answer your question?

**Reply by Coder-GAN** [reply]
Note that to optimize your workload, you can declare multiple measures in a single simulation, Eradiate will automatically simulate all of them when you will call the eradiate.run() function.

**Comment by leroyvn** [comment]
One important point: which mode are you using? `mono` or `ckd`?

**Reply by Coder-GAN** [reply]
I use the “mono” mode.

**Reply by leroyvn** [reply, marked as answer]
I think this is your problem: in `mono` mode, the post-processing pipeline does not include (yet) the SRF convolution. The root reason is that rigorously speaking, the convolution operation in monochromatic mode requires making assumptions on how to "fill" the spectral space between each spectral radiance record, and we haven't yet decided whether an automatic SRF convolution operation will be robust enough (in terms of how relevant the output can be). However, we plan to, at least, provide guidance on how to do this manually as part of post-processing.

Now, on how to solve your problem: if your spectral grid is dense enough, you can safely apply a convolution operation similar to what we do in the [`apply_spectral_response()`](https://github.com/eradiate/eradiate/blob/05fa0db5af9b53b3fa74f8f3ff31f0eed15e2493/src/eradiate/pipelines/logic.py#L246-L250) function.

A few more general comments:

* If you do not absolutely want to perform monochromatic simulations, you can also use the CKD mode, which will perform the SRF convolution safely.
* I would recommend using the `sentinel_2a-msi-XX` datasets rather than the `sentinel_2a-msi-XX-raw` for optimal performance.

**Reply by Coder-GAN** [reply]
Thank you so much for the suggestions!

According to your comments, I would like to use the CKD mode compared with the MONO mode. To be honest, I am not familiar with the difference between CKD and MONO mode, the reason I used the MONO mode is simply to follow the tutorials you provided.

Meanwhile, I would say that my purpose for the simulation was to compare the simulated (Eradiate) and monitored (Sentinel-2) BRF. In this case, is it ok or better to use CKD mode?

**Reply by leroyvn** [reply]
The `ckd` mode is a band mode: it discretizes the spectral domain in "bins". Without an atmosphere, it does not make much difference compared to a monochromatic simulation: one (or more, depending on the spectral quadrature) monochromatic simulation is run for each bin. The fundamental difference is that surface properties are assumed constant over each bin: this is where we make the assumption on "how to fill the spectral space between each spectral radiance record". I recommend reading [guide on spectral discretization](https://eradiate.readthedocs.io/en/stable/rst/user_guide/spectral_discretization.html) to understand how Eradiate behaves.

Which mode is "best" depends on the data you have and what you want to simulate. Generally, if you have an atmosphere in your scene, you should use the `ckd` mode. If you don't, it's up to you: monochromatic and CKD with 1 spectral quadrature point per bin are equivalent.

**Reply by Coder-GAN** [reply]
Thanks again for the explanation and suggestions, I will go to read the content you suggested and go through the behavior of Eradiate.

**Reply by Coder-GAN** [reply]
Hello @leroyvn, I am sorry to bother you again. I tried to declare multiple measures in a single simulation, the below are my codes:

        measure = [
            {
                "id": 'B4',
                "type": "mdistant",
                "construct": 'from_angles',
                "angles": [vza, vaa],
                "spp": 10000,
                "azimuth_convention": 'NORTH_LEFT',
                "srf": 'sentinel_2a-msi-4'
            },
            {
                "id": 'B5',
                "type": "mdistant",
                "construct": 'from_angles',
                "angles": [vza, vaa],
                "spp": 10000,
                "azimuth_convention": 'NORTH_LEFT',
                "srf": 'sentinel_2a-msi-5'
            },
            {
                "id": 'B6',
                "type": "mdistant",
                "construct": 'from_angles',
                "angles": [vza, vaa],
                "spp": 10000,
                "azimuth_convention": 'NORTH_LEFT',
                "srf": 'sentinel_2a-msi-6'
            },
            {
                "id": 'B7',
                "type": "mdistant",
                "construct": 'from_angles',
                "angles": [vza, vaa],
                "spp": 10000,
                "azimuth_convention": 'NORTH_LEFT',
                "srf": 'sentinel_2a-msi-7'
            }
        ]

After running "ds = eradiate.run(exp)",  the simulation actually runs successfully but gets the "ValueError: '690' is not in list",  is my declaration incorrect?

**Comment by leroyvn** [comment]
I cannot see how you access the data returned by [`eradiate.run()`](https://eradiate.readthedocs.io/en/stable/rst/reference_api/eradiate_core.html#eradiate.run), but I can already tell you one thing: this function will return a dictionary if you define multiple measures. It's a bit like automatically squeezing an array returned by a function. 

We are open to discussing this behaviour, arguably not very intuitive but very convenient in case of a single measure.
